gc/2022/GC-02-08.md

```diff
@@ -0,0 +1,266 @@
+
+ASC: ???
+
+CW: In a world with the union constructor, you wouldn’t use a cast(?) you would construct a union.
+
+BT: Union or type or heap type?
+
+CW: Heap type.
+
+BT: Why isn’t null a heap type?
+
+AR: You could make null a heap type. But it has this very special role.
+
+ASC: Having null as a separate type could be useful, that would keep us from having to specify a heap type for the null.
+
+AR: That would be separate, adding a bottom heap type, which couldn’t be anything but null.
+
+BT: Could it be a non-null bottom type?
+
+AR: It would be separate because bottom is a subtype of everything else, but null is disjoint from everything else.
+
+CW: My assumption is that enough languages will want to be using null and nullable types in such a way that it makes sense for null to be a dimension. i31 is much less generally-applicable.
+
+AR: Agree with CW that keeping null as a separate dimension makes sense given its special role. Even within Wasm null has a special role: e.g. default value of locals.
+
+TL: Makes sense that null is special. But this conversation is about whether i31 is special. Clearly it's somewhat special, given we don't have i63 or whatever. Why is it not special enough to be a type dimension?
+
+CW: I think it's special enough to merit a union constructor, but not a type dimension.
+
+AR: You could require usage of i31 to have a defined type; we don't do that because it's kind of stupid, but it would make sense.
+
+BT: Could declare structs to also be able to be i31. 
+
+CW: That’s what I was talking about with the special case union constructor.
+
+ASC: I don't think it will work well for a struct type to declare itself as being also-an-i31. Since the thing you want to deal with, when it's a struct, would be only-i31.
+
+AR: Right, you want a struct type and a union of defined-struct-type and i31.
+
+BT: But then how do we talk about only ref types that are pointers (not i31)?
+
+CW: That would be a union of data and func.
+
+AR: If we had unions we wouldn't need `any` or `eq`, they could be unions with i31.
+
+BT: I want to define any point in the space of pointer-only types and also any point in the space of tagged pointers. Want a type as specific as possible.
+
+AR: And a union would give you that.
+
+CW: I think the point BT was making earlier was that adding i31 as a dimension would imply the use of tagging.
+
+AR: Special union can’t use tagging either to make a representation change.
+
+BT: Regardless of whether it's a dimension or a possible feature of unions, you'd have to pull it out of the hierarchy.
+
+ASC: That would be necessary to avoid checking for i31 when downcasting from eq or any.
+
+AR: If you want to avoid those checks you can just avoid using `any` and `eq` (those are just unions).
+
+TL: AR, you said that you couldn't use pointer tagging with these unions? Doesn't that make i31 useless?
+
+CW: Downcasting would perform the check and could change the representation.
```

Comment:
```suggestion
CW: Pointer tagging should still work? Downcasting won't change representation, values of precise types will keep their tags.
```

This better captures what I said during the meeting, although I don't think this point was discussed further so I could have said something incorrect.

Reply:
Great, thanks for the edit!